### DISCUSSION TITLE
Public/private body for #[spec] functions

### DISCUSSION THREAD

**Discussion by utaal** [verus-lang/verus#56] [Language design]
We can currently control whether the axioms that defines the function body is visible in other modules with `#[verifier(pub_abstract)]`.

@tjhance suggests to use naming, and visibility control, more similar to rust's: "it would be nice if we could use the same language as rust, e.g., pub, pub(crate), etc."

As @Chris-Hawblitzel points out, keywords cannot appear inside attributes, unfortunately. We could use string literals instead (@tjhance).

@utaal suggests that default maybe should be `#[verifier(pub_abstract)]` and then one can expand visibility, like in rust. This also helps cut down on the z3 definitions by default, that is: it helps the developer control when things become transparent more directly. This is particularly helpful for e.g. an invariant function on a data structure, which should be "transparent" inside the module that defines the datatype, but always abstract to users.

Maybe something like this?

* `#[spec] fn inv() -> bool` is a private function
* `#[spec] pub fn inv() -> bool` is a public function which is defined within the module, but remains uninterpreted outside
* `#[spec] #[pub_spec] pub fn thing() -> bool` is a public function that published the function axiom to external modules 

Note: pub_abstract interacts with adt field pub: a function can only be pub and not-pub-abstract if it doesn’t reference any fields that aren’t pub. This is (also) because structs with private fields are encoded as z3 sorts, instead of datatypes. The limitation currently is, IIRC, that either all of the fields are pub or none are.

**Comment by utaal** [comment]
If we're making changes, I think we should also consider having "spec" in the attribute name, as this visibility control only makes sense for `#[spec]` functions (as `#[proof]` and `#[exec]` are not directly encoded.

**Comment by utaal** [comment]
As promised, I looked into this. Here's my proposed syntax, which is implemented here https://github.com/secure-foundations/verus/pull/65 

The following applies to *both* free-standing functions and associated functions (`impl`s), additional restrictions for the latter are explained later.

* `#[spec] fn foo(...) -> ...` is private to its module, and non-opaque by default within the module;
* `#[spec] #[verifier(opaque)] fn foo(...) -> ...` is private to its module, and opaque within the module;
* `#[spec] pub fn foo(...) -> ...` is non-opaque within the module, and always uninterpreted outside its module;
* `#[spec] #[verifier(opaque)] pub fn foo(...) -> ...` is opaque within the module, and always uninterpreted outside its module;
* `#[spec] #[verifier(publish_opaque)] pub fn foo(...) -> ...` is non-opaque within the module, and opaque outside the module;
* `#[spec] #[verifier(publish)] pub fn foo(...) -> ...` is non-opaque within the module and outside the module; 
* `#[spec] #[verifer(opaque, publish)] pub fn foo(...) -> ...` is opaque within the module and outside the module;
* `#[spec] #[verifier(opaque, publish_opaque)] pub fn foo(...) -> ...` is meaningless;

As before, one can `reveal(...)` a function that is opaque within the same module or published and opaque.
Functions that are not uninterpreted outside the module (i.e. they are `publish` or `publish_opaque`) cannot refer to private fields or functions.

**Reply by tjhance** [reply]
To be honest, I... find this incredibly confusing.

 * The fact that `opaque, publish` and `publish_opaque` are different things is undoubtedly going to cause confusion
 * There isn't a clear relationship to Rust's visibility levels, e.g., there isn't a knob for the equivalent of `pub(crate)` and I'm not sure how this proposal would be cleanly extended to include such
   * What if I want something transparent to the crate, and published opaquely outside the crate?

I think part of the complexity of the proposal comes from the fact that `publish` is a "positive" property while `opaque` is a "negative" property, and I think it might be clearer to use "transparent" instead of "opaque", except in the case of declaring something completely opaque

so for example we could do:

`publish` - what you called `publish_opaque`
`publish, transparent` (or `publish_transparent`) - what you called `publish`
`publish(...), transparent(...)` for any possible combination
`publish, opaque` - what you called `publish, opaque`

**Reply by tjhance** [reply]
BTW, I like the word "publish". It's cute & gets around the "pub" keyword problem but sounds similar.

**Reply by jaybosamiya** [reply]
How about the following alternative naming scheme?

`pub` should behave like Rust (i.e., it talks about whether the signature is even visible externally), and is behavior is orthogonal to the `#[verifier(...)]` declarations. If any particular combination is seen to be meaningless, then we hard-error.

The default is to have transparent internal to a module, and uninterpreted outside a module. To modulate away from the defaults, you can use the table below.

|                | Unmodified    | `opaque`   | `pub_revealable` | `pub_transparent` |
|----------------|---------------|------------|------------------|-------------------|
| Inside module  | Transparent   | Opaque     |                  | Transparent       |
| Outside module | Uninterpreted |            | Opaque           | Transparent       |

Using `pub_transparent` along with either of the other two is an immediate hard-error.


The reasoning with the above is the following:

+ The defaults keep things scoped to within the module but still easy to use. It also cannot "leak" outside the module without explicitly saying _something_ (note that this default is same as the original default by @utaal, no changes here)

+ `opaque` only modulates opaqueness within the module. Since the default within a module is transparent, this _is_ a restricting action, so should have a "negative" name to it.

+ `pub_revealable` only modulates opaqueness outside the module. Specifically, it gives the outside world the ability to reveal from opaqueness. Since this is derestricting the outside world, it is a "positive" property; however it is not _completely_ derestricting, but only giving some additional ability to the outside world, and thus it is `pub_revealable` and not `pub_reveal` or similar.

+ `pub_transparent` only modules outside the module, but I mark it as impacting both inside and outside in the above table, since it doesn't make sense whatsoever to have something be opaque internally but transparent externally. Anyways, it is an even stronger derestriction to the outside world, it uses a word that specifies that the action is _done_ rather than "could be done".

+ Having everything that impacts the outside world be marked with a `pub_` makes it extremely clear that it impacts strictly-outside-module stuff. Additionally it allows for forwards compatibility with situations where we may want to introduce something like `pub(crate)`/`pub(super)` but for opaqueness, by simply making it `pub_revealable(crate)`/...

**Reply by tjhance** [reply]
I really like `pub_revealable`, because it gets around a problem with an issue I had with both my and andrea's proposals: namely,
`publish` isn't explicit enough, while
`publish_opaque` is confusing due to name overlap.

`revealable` conveys the concept perfectly (it's not revealed right now, but it could be!) while avoiding the name overlap problem.

**Reply by tjhance** [reply]
I think Jay's proposal is very close to andrea's with the mapping

publish -> pub_transparent
publish_opaque -> pub_revealable

the difference is that andrea's proposal allows `publish, opaque` but not `publish_opaque, opaque` whereas
jay's proposal is the other way: it allows `pub_revealable, opaque` but not `pub_transparent, opaque` (since as jay says, it doesn't make sense for something to be transparent publicly but opaque inside the module)

**Reply by utaal** [reply]
@jaybosamiya can you confirm that 

> Jay's proposal is very close to andrea's with the mapping
> publish -> pub_transparent
> publish_opaque -> pub_revealable

where the allowed combinations of opaque/transparent are the same, but the naming changes?

In other words, I think we get to the same naming scheme, with the above replacement and by making this change to my suggestion:

* `#[spec] #[verifer(opaque, publish_opaque)] pub fn foo(...) -> ...` is opaque within the module and outside the module;
* `#[spec] #[verifier(opaque, publish)] pub fn foo(...) -> ...` is meaningless;

**Comment by jonhnet** [comment]
Many degrees of freedom here, but I like where this is converging: I agree that "revealable" is a goldilocks adjective. @jaybosamiya's tabular presentation makes a lot of sense to me; it clarifies why we need a restricting modifier inside the module and relaxing modifiers outside.

I _think_ I understand the design space, but I reserve the right to be confused again later. :v)

**Reply by jonhnet** [reply]
Suggestion: fill in the empty boxes in jay's table with "(still transparent)" and "(still uninterpreted)".

**Comment by Chris-Hawblitzel** [comment]
I'm assuming that we think that it will be common and useful to make a spec function body non-opaque privately and visible, but opaque, publicly.  Otherwise, we wouldn't bother to bring opaque-ness into this discussion in the first place.  After all, it would be simpler if opaqueness had nothing to do with modules and private/public.  But maybe we can go ahead with the extra complication and see how useful it turns out to be.  If it's useful and not too confusing, great.  If not, we can remove the feature.

It's potentially confusing to have a function behave differently in proofs depending on which module the proof is in.  The same proof in two different modules could succeed or fail depending on which modules the function appears opaque to, and, in contrast to ordinary public/private visibility, you wouldn't get a clear error message, just an SMT failure.  So I think we need to be very explicit about when this might happen.  In particular, I prefer something like `publish_opaque` (explicitly signaling the surprising behavior) to just `publish` for this case.

To make the concepts more discoverable for newcomers reading the code, I think we should try to minimize the number of different names for the same concepts.  For example, `revealable` sounds like a synonym for `opaque`.  I don't think we should have both.  Likewise, `transparent` is a name for "not `opaque`" -- do we need both names?

We could try `pub_reveal` (or `pub_revealed`) and `pub_opaque`.  That way users only have to know about two words, opaque and reveal.  And if we want to get even more elaborate, we can say things like `pub_opaque` together with `pub_reveal(crate)`.

**Comment by utaal** [comment]
Let's break this down.

* the first change here is to make `#[spec] pub` functions non-interpreted by default outside the module. It sounds like we agree on this;

* we know that `#[opaque]` is useful, and that the default (at least for private functions) should be non-opaque;

* like Chris, I think it would be confusing to have multiple names for the same concept, thus I don't think we should introduce `revealable` or similar names; if we want to control opaque-ness outside the module, we should use something with `opaque` in the name.

* I don't think matching all the rust visibility levels should be a goal: we don't know for sure whether controlling opaque-ness outside the module separately is useful yet, so I think it's too early to design for even more granular controll of opaque-ness.

`publish` and `opaque` I think combine well: `publish` controls whether something is (un)interpreted outside the module, and `opaque` controls the other axis.

I don't particularly like `pub_reveal`/`pub_revealed` because it conflates the (un)interpreted axis with the opaque-ness axis.

---

Here's my revised proposal, paraphrasing Jay's table. When two modifiers appear in the same column, one needs to apply both. This is the same allowed combinations as my original proposal, with the names revised.

|                | Unmodified    | `opaque`   | `publish` | `publish` `opaque` | `publish` `pub_opaque` |
|----------------|---------------|------------|------------------|-------------------|------------------|
| Inside module  | Transparent   | Opaque     | Transparent | Opaque       | Transparent |
| Outside module | Uninterpreted | Uninterpreted | Transparent           | Opaque       | Opaque |

Unmodified, `opaque`, and `publish` are likely the most common, and require fewer modifiers.

`publish` only controls whether something is (un)interpreted outside the module.

`opaque` operates as before, and `pub_opaque` only applies opacity outside the module. This explicitly signals the surprising behaviour and makes removing this additional feature easiest, if we decide we don't need it.

**Reply by jonhnet** [reply]
Is the axis-conflation concern that "pub" is the Rust word for the name appearing to outside clients? That makes sense. That suggests "pub_opaque" should be "publish_opaque"? (Or better yet "publish_opaquely".)

So the way to document this revised proposal:
* By default, function definitions are transparent inside a Rust module (for great automation) and opaque from outside (to limit SMT clutter at what should already be an interface).
* The basic "opaque" modifier reduces automation by making the definition opaque (but revealable) inside the module.
* The basic "publish" modifier adds outside automation by publishing the function definition to clients.
* Publish+opaque means "first publish this definition outside, but then opaque both inside and outside until revealed"
* publish + publish_opaque means "first publish this definition outside, but then opaque the outside definition until revealed"

Telling the story like that, I think I understand the reasoning behind the kinda-unexpected interactions among those knobs.

I'm wondering if we like the term "opaque" because we've been using it so long. :v) To a new reader, "opaque" and "uninterpreted" both sound like the opposite of "transparent." The salient feature of "opaque" is that it's kinda non-transparent but controllably-so. For your consideration: rename "opaque" to "revealable" everywhere: if a function is "revealable", it's evidently not transparent yet, but evidently can be made so.

**Comment by Chris-Hawblitzel** [comment]
How about `publish_as_opaque` rather than `publish_opaquely`, since the end result is opaque (an adjective), rather than the act of publishing being opaque (an adverb)?

I'd prefer "opaque" to "revealable" because "opaque" conveys the primary thing that you need to know: that you can't see something.  It's why I'd rather say "locked" than "unlockable" or "wet" than "dryable"; I want to first say what something is, and only later say what might change it.

**Comment by utaal** [comment]
I'm going to make an executive decision here, as the only remaining concern seems to be the name for the `pub_opaque` concept.

* Uninterpreted / `publish` will be the axis that controls whether a function is (un)interpreted outside the module, and
* Transparent / `opaque` / `opaque_outside_module` will be the axis that controls where it is opaque/revealed.
